File: src/main/java/edu/cornell/cs/apl/viaduct/algebra/HeytingAlgebra.java
package edu.cornell.cs.apl.viaduct.algebra;

/**
 * A Heyting algebra is a bounded lattice that supports an implication operation {@code →} where
 * {@code A → B} is the greatest element {@code x} that satisfies {@code A ∧ x ≤ B}.
 *
 * <p>This is also called an implicated lattice.
 *
 * @see <a href="https://en.wikipedia.org/wiki/Pseudocomplement#Relative_pseudocomplement">Wikipedia
 *     page on pseudocomplement</a>
 */
public interface HeytingAlgebra<T extends HeytingAlgebra<T>> extends Lattice<T> {
  /** {@code this.imply(that)} is the greatest solution to {@code this.meet(x) ≤ that}. */
  T imply(T that);
}

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/algebra/HeytingAlgebra.kt
package edu.cornell.cs.apl.viaduct.algebra

/**
 * A Heyting algebra is a bounded lattice that supports an implication operation `→` where
 * `A → B` is the greatest element `x` that satisfies `A ∧ x ≤ B`.
 *
 * This is also called an implicated lattice.
 *
 * @see [Wikipedia page on pseudocomplement](https://en.wikipedia.org/wiki/Pseudocomplement.Relative_pseudocomplement)
 */
interface HeytingAlgebra<T : HeytingAlgebra<T>> : Lattice<T> {
    /** `this.imply(that)` is the greatest solution to `this.meet(x) ≤ that`. */
    fun imply(that: T): T
}
